README.md
# M2mKeygen

This gem exists for simplifying Machine to Machine signature generation and verification in a secure way.

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ yarn add m2m-keygen

or if using npm

    $ npm install --save m2m-keygen

## Usage

### Signature

This gem provides a module for signing and checking signature for HTTP requests

Every method will use your secret key and eventually an encryption algorithm.

```ts
import { sign } from 'm2m-keygen';

const signed = sign({ secret: "my_secret_key", algorithm: "sha256", ... });

const signed = sign({ secret: "my_secret_key", ... }); // => Will default algorithm to sha512
```

#### Signing

Use the `sign` method to generate a new signature.

- `params` is a params hash. The order of keys isn't important as the lib will reformat them.
- `verb` is the http verb
- `path` is the path for the request

```ts
import { sign } from "m2m-keygen";

sign({
  secret: "my_secret_key",
  params: {
    a: "test",
    b: 1,
    d: ["a", "b"],
    c: {
      e: 45,
    },
  },
  verb: "get",
  path: "/path",
}); // => "a52168521868ebb37a38f90ec943163d9acb6ceb982206f437e1feb9ca32e7c1a8edef68f0ff4e195aeca1da93ae9afc8da214cb51a812fc6cc3730fdc7613fa"
```

After generating the signature send it alongside your request for verification on the receiver side.

#### Verifying

Use the `validate` method to verify that a received signature correspond to the HTTP request.

- `params` is a params hash. The order of keys isn't important as the lib will reformat them.
- `verb` is the http verb
- `path` is the path for the request
- `signature` is the received signature

```ts
import { validate } from "m2m-keygen";

validate(
  secret: "my_secret_key",
  params: {
    a: "test",
    b: 1,
    d: ["a", "b"],
    c: {
      e: 45,
    },
  },
  verb: "get",
  path: "/path",
  signature:
    "a52168521868ebb37a38f90ec943163d9acb6ceb982206f437e1feb9ca32e7c1a8edef68f0ff4e195aeca1da93ae9afc8da214cb51a812fc6cc3730fdc7613fa"
) // => true
```

If the validation is true, the request was signed with the same algorithm and same secret key.

#### Fetch API with auto signature generation and expiry

An helper as been added to generate a fetch function that will automatically sign the request and add an expiry.

The `fetcher` argument is any function that is compatible with the fetch API.

The result behaves slightly differently than a normal fetch function. It expects you to use 3 arguments:

- The first one is the path without any query arguments
- The second one is an object with the params to be sent
- The last one is anything you would like to pass to the fetch function as a second argument (method, headers, …).

The fetcher will add an expiry date to 90 seconds in the future (unless a previous expiry is already in params).
It will then transform the params into query arguments or encode them as JSON in the body depending on the method (GET, POST, …) you use.

```ts
import { generateFetcher } from "m2m-keygen";

const fetcher = generateFetcher({
  fetcher: fetch, // fetch can be any function compatible with the fetch API
  secret: "secret",
  algorithm: "sha512", // optional, default to sha512
  headerName: "X-Signature", // optional, default to X-Signature
});

// The generated fetcher, except for the params/body/query will behave exactly like fetch does.
fetcher(
  "http://example.com/oki",
  { a: 1, b: 2 },
  { method: "post", headers: { "My-Header": "Yay" } }
)
  .then((res) => {
    console.log(res);
  })
  .catch((err) => {
    console.log(err);
  });
```

#### Helpers

This lib exposes a few helpers that can be used in other projects.

##### translateParamsToQuery

This helper will translate a params hash to a query string.

```ts
import { translateParamsToQuery } from "m2m-keygen";

translateParamsToQuery({
  a: 1,
  b: "2",
  c: true,
  d: false,
  e: [1, 2, 3],
  f: { a: 1, b: 2 },
  g: null,
}); // => '?a=1&b=2&c=true&d=false&e[]=1&e[]=2&e[]=3&f[a]=1&f[b]=2&g=false'
```

##### secureCompare

This helper will compare two strings in a secure way.
Extracted from: [vadimdemedes/secure-compare](https://github.com/vadimdemedes/secure-compare)

```ts
import { secureCompare } from "m2m-keygen";

secureCompare("abc", "abc"); // => true

secureCompare("你好世界", "你好世界"); // => true

secureCompare("abc", "ab"); // => false

secureCompare("abc", "abd"); // => false
```

## How does it works

This is intended for a secure discussion between 2 servers and not something in a browser as the secret key must be stored and used both side (and you don't want to send the secret key in the browser).

Both server will have the same secret key.
The sender will generate a signature matching the HTTP request it will be sending and add it to the request in a designated header.
The receiver will generate the same signature from the HTTP request it has received and will compare it with the signature in the header.

The comparison will be done in constant time (i.e. secure) because both string will be hexdigest from a HMAC with the same algorithm.

## Development

After checking out the repo, run `yarn install` to install dependencies. Then, run `yarn test` to run the tests.

Every commit/push is checked by husky.

Tool used in dev:

- ESlint
- Prettier
- TypeScript
- Jest

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/BillCorporate/m2m_keygen_ts. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/BillCorporate/m2m_keygen_ts/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the M2mKeygen project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/BillCorporate/m2m_keygen_ts/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2033-10-04

### Changed

- Update dependencies

## [1.0.0] - 2022-09-01

### Added

- `generateFetcher` function have been added.
- Helpers functions have been exposed.

### Changed

- Signature container has been removed and all methods are flattened.

## [0.2.0] - 2022-08-31

### Added

- `sign` and `validate` functions in Signature object

### Changed

- Activating cache for GH actions

## [0.1.3] - 2022-08-31

### Changed

- Added various fields in package.json

## [0.1.2] - 2022-08-31

### Added

- Proper Readme
- Husky and Lint-Staged

## [0.1.1] - 2022-08-30

### Changed

- Fixing deploy

## [0.1.0] - 2022-08-30

### Added

- Basic skeleton for the package

[unreleased]: https://github.com/Billcorporate/m2m_keygen_ts/compare/v1.0.1...HEAD
[1.0.1]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v1.0.1
[1.0.0]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v1.0.0
[0.2.0]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v0.2.0
[0.1.3]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v0.1.3
[0.1.2]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v0.1.2
[0.1.1]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v0.1.1
[0.1.0]: https://github.com/Billcorporate/m2m_keygen_ts/releases/tag/v0.1.0


tests/index.test.ts
import * as All from '../src/index';

describe('Signature', () => {
  it('exposes sign', () => {
    expect(All.sign).toBeDefined();
  });

  it('exposes validate', () => {
    expect(All.validate).toBeDefined();
  });

  it('exposes secureCompare', () => {
    expect(All.secureCompare).toBeDefined();
  });

  it('exposes generateFetcher', () => {
    expect(All.generateFetcher).toBeDefined();
  });

  it('exposes translateParamsToQuery', () => {
    expect(All.translateParamsToQuery).toBeDefined();
  });
});


tests/fetcher.test.ts
import { sign } from '../src/signature';
import { generateFetcher } from '../src/fetcher';
import { mockDate } from './support/mockDate';

// Create a mock fetcher
const mockFetch = jest.fn((url, options) => ({
  result: 'ok',
}));

describe('generateFetcher', () => {
  const frozenTime = new Date();
  const timestamp = Math.round(frozenTime.getTime() / 1000);

  let unfreezeTime: Function;

  beforeAll(() => {
    unfreezeTime = mockDate(frozenTime);
  });

  afterAll(() => {
    unfreezeTime();
  });

  const fetcher = generateFetcher({
    fetcher: mockFetch,
    secret: 'secret',
  });

  it('generates a fetcher', () => {
    fetcher('http://example.com/oki', { a: 1, b: 2 });

    expect(mockFetch).toHaveBeenCalled();
    expect((mockFetch.mock.lastCall || [])[0]).toEqual(
      `http://example.com/oki?a=1&b=2&expiry=${timestamp + 90}`
    );
    expect((mockFetch.mock.lastCall || [])[1].headers['X-Signature']).toEqual(
      sign({
        secret: 'secret',
        params: { a: 1, b: 2, expiry: timestamp + 90 },
        verb: 'GET',
        path: '/oki',
      })
    );
    expect((mockFetch.mock.lastCall || [])[1].headers['Accept']).toEqual(
      'application/json'
    );
  });

  it('keeps previous expiry', () => {
    fetcher('http://example.com/oki', { a: 1, b: 2, expiry: timestamp + 100 });

    expect(mockFetch).toHaveBeenCalled();
    expect((mockFetch.mock.lastCall || [])[0]).toEqual(
      `http://example.com/oki?a=1&b=2&expiry=${timestamp + 100}`
    );
    expect((mockFetch.mock.lastCall || [])[1].headers['X-Signature']).toEqual(
      sign({
        secret: 'secret',
        params: { a: 1, b: 2, expiry: timestamp + 100 },
        verb: 'GET',
        path: '/oki',
      })
    );
  });

  it('keeps existing headers', () => {
    fetcher(
      'http://example.com/oki',
      { a: 1, b: 2 },
      { headers: { Accept: 'text/plain', Pokemon: 'Red' } }
    );

    expect(mockFetch).toHaveBeenCalled();
    expect((mockFetch.mock.lastCall || [])[1].headers['Accept']).toEqual(
      'text/plain'
    );
    expect((mockFetch.mock.lastCall || [])[1].headers['Pokemon']).toEqual(
      'Red'
    );
  });

  describe('when the method is specified', () => {
    it('uses the method in signature', () => {
      fetcher('http://example.com/oki', { a: 1, b: 2 }, { method: 'post' });

      expect(mockFetch).toHaveBeenCalled();
      expect((mockFetch.mock.lastCall || [])[0]).toEqual(
        `http://example.com/oki`
      );
      expect((mockFetch.mock.lastCall || [])[1].headers['X-Signature']).toEqual(
        sign({
          secret: 'secret',
          params: { a: 1, b: 2, expiry: timestamp + 90 },
          verb: 'POST',
          path: '/oki',
        })
      );
    });

    it('keeps the method in the fetch options', () => {
      fetcher('http://example.com/oki', { a: 1, b: 2 }, { method: 'post' });

      expect(mockFetch).toHaveBeenCalled();
      expect((mockFetch.mock.lastCall || [])[0]).toEqual(
        `http://example.com/oki`
      );
      expect((mockFetch.mock.lastCall || [])[1].method).toEqual('post');
    });

    it('JSON encodes the params in the body', () => {
      fetcher('http://example.com/oki', { a: 1, b: 2 }, { method: 'post' });

      expect(mockFetch).toHaveBeenCalled();
      expect((mockFetch.mock.lastCall || [])[0]).toEqual(
        `http://example.com/oki`
      );
      expect((mockFetch.mock.lastCall || [])[1].body).toEqual(
        JSON.stringify({ a: 1, b: 2, expiry: timestamp + 90 })
      );
    });

    it('tells the app the content is JSON', () => {
      fetcher('http://example.com/oki', { a: 1, b: 2 }, { method: 'post' });

      expect(mockFetch).toHaveBeenCalled();
      expect((mockFetch.mock.lastCall || [])[0]).toEqual(
        `http://example.com/oki`
      );
      expect(
        (mockFetch.mock.lastCall || [])[1].headers['Content-Type']
      ).toEqual('application/json');
    });
  });
});


tests/signature.test.ts
import { sign, validate, secureCompare } from '../src/signature';

describe('#sign', () => {
  const secret = 'secret';
  const algorithm = 'sha256';
  const verb = 'get';
  const path = '/path';
  const params = { b: '1', a: '2' };

  it('returns signature', () => {
    expect(sign({ secret, algorithm, params, verb, path })).toEqual(
      // Signature generated using OpenSSL cli
      // echo "GET/patha=2&b=1" | perl -0 -pe 's/\n\Z//' | openssl sha256 -hmac "secret"
      '7d86fa8a7f871589697b0f41542d065b1ddbbe155e83349fffd937edcfa85af7'
    );
  });
});

describe('#validate', () => {
  const secret = 'secret';
  const algorithm = 'sha256';
  const verb = 'get';
  const path = '/path';
  const params = { b: '1', a: '2' };

  describe('with a valid signature', () => {
    const signature =
      '7d86fa8a7f871589697b0f41542d065b1ddbbe155e83349fffd937edcfa85af7';

    it('returns true', () => {
      expect(
        validate({ secret, algorithm, params, verb, path, signature })
      ).toBe(true);
    });
  });
  describe('with a invalid length signature', () => {
    const signature =
      '7d86fa8a7f871589697b0f41542d065b1ddbbe155e83349fffd937edcfa85af';

    it('returns false', () => {
      expect(
        validate({ secret, algorithm, params, verb, path, signature })
      ).toBe(false);
    });
  });
  describe('with a invalid signature', () => {
    const signature =
      '7d86fa8a7f871589697b0f41542d065b1ddbbe155e83349fffd937edcfa85af8';

    it('returns false', () => {
      expect(
        validate({ secret, algorithm, params, verb, path, signature })
      ).toBe(false);
    });
  });
});

describe('secureCompare', () => {
  it('return true if the strings are identical', () => {
    expect(secureCompare('abc', 'abc')).toBe(true);
  });

  it('return true if the strings are identical in utf8', () => {
    expect(secureCompare('你好世界', '你好世界')).toBe(true);
  });

  it('should return false if the strings are different lengths', () => {
    expect(secureCompare('abc', 'ab')).toBe(false);
  });

  it('should return false if the strings have different contents', () => {
    expect(secureCompare('abc', 'abd')).toBe(false);
  });
});


src/signature.ts
import { createHmac } from 'crypto';
import { encodes } from './helpers/paramsEncoder';
import { ParamsTypes } from './types/paramsTypes';

/**
 * Securely compare two strings
 * Extracted from : https://github.com/vadimdemedes/secure-compare
 *
 * @param {string} a - first string
 * @param {string} b - second string
 * @returns {boolean} - true if strings are equal, false otherwise
 */
export const secureCompare = (a: string, b: string): boolean => {
  if (typeof a !== 'string' || typeof b !== 'string') return false;

  var mismatch = a.length === b.length ? 0 : 1;
  if (mismatch) {
    b = a;
  }

  for (var i = 0, il = a.length; i < il; ++i) {
    mismatch |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }

  return mismatch === 0;
};

/**
 * Generate a signature from the different part of the request.
 *
 * @param {string} secret - The secret used to sign the request.
 * @param {string|undefined} algorithm - algorithm to use - Default to sha512
 * @param {ParamType} params - params to sign
 * @param {string} verb - http verb
 * @param {string} path - path to sign
 * @returns {string} - The generated signature
 */
export const sign = ({
  secret,
  algorithm = 'sha512',
  params,
  verb,
  path,
}: {
  secret: string;
  algorithm?: string;
  params: ParamsTypes;
  verb: string;
  path: string;
}): string => {
  const hmac = createHmac(algorithm, secret);

  hmac.update(`${verb.toUpperCase()}${path}${encodes(params)}`);

  return hmac.digest('hex');
};

/**
 * Validate the signature of a request.
 *
 * @param {string} secret - The secret used to sign the request.
 * @param {string|undefined} algorithm - algorithm to use - Default to sha512
 * @param {ParamType} params - params to sign
 * @param {string} verb - http verb
 * @param {string} path - path to sign
 * @param {string} signature - signature to check against
 * @returns {boolean} - Is signature is valid
 */
export const validate = ({
  secret,
  algorithm = 'sha512',
  params,
  verb,
  path,
  signature,
}: {
  secret: string;
  algorithm?: string;
  params: ParamsTypes;
  verb: string;
  path: string;
  signature: string;
}) => {
  return secureCompare(
    sign({ secret, algorithm, params, verb, path }),
    signature
  );
};


src/helpers/paramsEncoder.ts
import { ParamsTypes, ParamsValueType } from '../types/paramsTypes';

type FetchParamsType = {
  [key: string]:
    | null
    | boolean
    | number
    | string
    | Array<boolean | number | string>
    | { [key: string]: boolean | number | string };
};

/**
 * Encode the parameters to be signed, Rails way.
 *
 * @param {FetchParamsType} params - params to sign
 * @returns {string} - The encoded parameters
 */
export const translateParamsToQuery = (params: FetchParamsType) => {
  if (Object.keys(params).length === 0) return '';

  return `?${Object.entries(params)
    .reduce<Array<string>>((current, [nextKey, nextItem]) => {
      let value: Array<string> = [];
      switch (typeof nextItem) {
        case 'boolean':
          value = [`${nextKey}=${nextItem ? 'true' : 'false'}`];
          break;

        case 'number':
          value = [`${nextKey}=${nextItem}`];
          break;

        case 'string':
          value = [`${nextKey}=${nextItem}`];
          break;

        case 'object':
          if (nextItem === null) {
            value = [`${nextKey}=false`];
          } else if (Array.isArray(nextItem)) {
            value = nextItem.reduce<Array<string>>(
              (currStrArr, nextArrValue) => [
                ...currStrArr,
                `${nextKey}[]=${nextArrValue}`,
              ],
              []
            );
          } else {
            value = Object.entries(nextItem).reduce<Array<string>>(
              (currStrHash, [nextHashKey, nextHashValue]) => [
                ...currStrHash,
                `${nextKey}[${nextHashKey}]=${nextHashValue}`,
              ],
              []
            );
          }
          break;

        default:
          break;
      }
      return [...current, ...value];
    }, [])
    .join('&')}`;
};

const jsonifyValue = (value: ParamsValueType): string | number | boolean => {
  if (typeof value !== 'object') return value;
  return JSON.stringify(value);
};

const encodeValue = (value: ParamsValueType): ParamsValueType => {
  if (typeof value === 'object' && !Array.isArray(value))
    return encodeHashValue(value);
  return value;
};

const encodeHashValue = (value: ParamsValueType): ParamsValueType => {
  return Object.fromEntries(
    Object.entries(value)
      .filter(
        ([_, value]) => value !== undefined && value !== null && value !== ''
      )
      .sort(([keyA, _vA], [keyB, _vB]) => keyA.localeCompare(keyB))
      .map(([key, value]) => [key, encodeValue(value)])
  );
};

export const encodes = (params: ParamsTypes): string => {
  if (params === null || params === undefined) return '';

  return Object.entries(params)
    .filter(
      ([_, value]) => value !== undefined && value !== null && value !== ''
    )
    .sort(([keyA, _vA], [keyB, _vB]) => keyA.localeCompare(keyB))
    .map(
      ([key, value]) =>
        `${key}=${jsonifyValue(encodeValue(value as ParamsValueType))}`
    )
    .join('&');
};


src/fetcher.ts
import { translateParamsToQuery } from './helpers/paramsEncoder';
import { sign } from './signature';

export const generateFetcher =
  ({
    fetcher,
    secret,
    algorithm = 'sha512',
    headerName = 'X-Signature',
  }: {
    fetcher: (entry: any, init?: any) => any;
    secret: string;
    algorithm?: string;
    headerName?: string;
  }) =>
  (entry: string, params: any = {}, init?: any) => {
    if (!init) init = {};
    const headers: { [key: string]: string } = { ...init.headers };
    const path = entry.replace(/^(.*\/\/)?[^\/]+/, '');

    params.expiry ||= Math.round(Number(new Date()) / 1000) + 90;

    headers[headerName] = sign({
      secret,
      algorithm,
      params: params,
      verb: init.method?.toLocaleUpperCase() || 'GET',
      path,
    });

    headers['Accept'] ||= 'application/json';

    if (init.method && init.method.toLocaleUpperCase() !== 'GET') {
      headers['Content-Type'] = 'application/json';
      init.body = JSON.stringify(params);
    } else {
      entry = `${entry}${translateParamsToQuery(params)}`;
    }

    init.headers = headers;

    return fetcher(entry, init);
  };
